### PR TITLE
Lw/api review v4

### PR DESCRIPTION
**Pull request by phi-lira**
### Purpose of this PR
 - Renamed/Refactored C# and Shader API to get out of experimental.
 - Implemented @andrem shader feedback requests
 - Improved shader stripping.

---
## Added
 - [Shader API] `GetMainLight` and `GetAdditionalLight` functions can now compute shadow attenuation and store it in the new `shadowAttenuation` field in `LightData` struct.
- [Shader API] Added `VertexPosition` struct that contains vertex position in difference spaces (world, view, hclip).
 - [Shader API] Added `GetVertexPosition` function to get an initialized `VertexPosition`.
 - [Shader API] Added `GetPerObjectLightIndex` function to return the per-object index given a for-loop index.
 - [Shader API] Added `GetShadowCoord` function that takes a `VertexPosition` as input.
 - Autodesk interactive shaders. 
## Changed
 - Reorganized LWRP asset settings to be more clear.
 - Vertex lighting now controls if additional lights should be shaded per-vertex or per-pixel.
 - Renamed all `Local Lights` nomenclature to `Additional Lights`.
 - Changed shader naming to conform to our SRP shader code convention.
 - [Shader API] Renamed `SpotAttenuation` function to `AngleAttenuation`.
 - [Shader API] Renamed `_SHADOWS_ENABLED` keyword to `_MAIN_LIGHT_SHADOWS`
 - [Shader API] Renamed `_SHADOWS_CASCADE` keyword to `_MAIN_LIGHT_SHADOWS_CASCADE`
 - [Shader API] Renamed `_VERTEX_LIGHTS` and `_ADDITIONAL_LIGHTS` keywords to
`_ADDITIONAL_LIGHTS_VERTEX` and `_ADDITIONAL_LIGHTS_PIXEL` respectively.
 - [Shader API] Renamed `_LOCAL_SHADOWS_ENABLED` to `_ADDITIONAL_LIGHT_SHADOWS`
 - [Shader API] Renamed `GetLight` function to `GetAdditionalLight`.
  - [Shader API] Renamed `GetPixelLightCount` function to `GetAdditionalLightsCount`.
  - [Shader API] Renamed `attenuation` to `distanceAttenuation` in `LightData`.
  - [Shader API] Renamed `GetLocalLightShadowStrength` function to `GetAdditionalLightShadowStrength`.
  - [Shader API] Renamed `SampleScreenSpaceShadowMap` functions to `SampleScreenSpaceShadowmap`.
  - [Shader API] Renamed `MainLightRealtimeShadowAttenuation` function to `MainLightRealtimeShadow`.
  - [Shader API] Renamed light constants from `Directional` and `Local` to `MainLight` and `AdditionalLights`.
  - [Shader API] Renamed `GetLocalLightShadowSamplingData` function to `GetAdditionalLightShadowSamplingData`.
  - [Shader API] Removed OUTPUT_NORMAL macro.
  - [Shader API] Removed `lightIndex` and `substractiveAttenuation` from `LightData`.
  - [Shader API] Removed `ComputeShadowCoord` function. `GetShadowCoord` is provided instead.

  
---
### Testing status
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?automation-tools_branch=add-platform-filter&ScriptableRenderLoop_branch=lw%2Fapi-review-v4&unity_branch=trunk

---
### Overall Product Risks
**Technical Risk**: High. Lot's of refactors and a few things changed.

---

### CODE REVIEW

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/DepthCopy.hlsl [hunk 1]
```diff
@@ -3,30 +3,30 @@
...
 {
     float4 vertex   : POSITION;
     float2 uv       : TEXCOORD0;
     UNITY_VERTEX_INPUT_INSTANCE_ID
 };
 
-struct VertexOutput
+struct Varyings
 {
     float4 position : SV_POSITION;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/DepthCopy.hlsl]: positionCS as in shaders below?

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/LightweightPassMetaCommon.hlsl [hunk 2]
```diff
@@ -36,7 +36,7 @@ struct MetaVertexInput
 #endif
 };
 
-struct MetaVertexOuput
+struct Varyings
 {
     float4 pos      : SV_POSITION;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/LightweightPassMetaCommon.hlsl]: Is this one worth changing to positionWS as well? :)

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Particles.hlsl [hunk 3]
```diff
@@ -164,20 +162,20 @@ struct VertexOutputLit
     float4 clipPos                  : SV_POSITION;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Particles.hlsl]: positionCS?

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Particles.hlsl [hunk 4]
```diff
@@ -136,16 +136,14 @@ struct appdata_particles
...
 };
 
-struct VertexOutputLit
+struct VaryingsParticle
 {
     half4 color                     : COLOR;
     float2 texcoord                 : TEXCOORD0;
-#if _NORMALMAP
+#ifdef _NORMALMAP
     half3 tangent                   : TEXCOORD1;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Particles.hlsl]: WS? all of them, I suppose?

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Shadows.hlsl [hunk 5]
```diff
@@ -160,58 +160,61 @@ real SampleShadowmap(float4 shadowCoord, TEXTURE2D_SHADOW_ARGS(ShadowMap, sample
...
+    half4 weights = half4(distances2 < _CascadeShadowSplitSphereRadii);
     weights.yzw = saturate(weights.yzw - weights.xyz);
 
     return 4 - dot(weights, half4(4, 3, 2, 1));
 }
 
 float4 TransformWorldToShadowCoord(float3 positionWS)
 {
-#ifdef _SHADOWS_CASCADE
+#ifdef _DIRECTIONAL_SHADOWS_CASCADE
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Shadows.hlsl]: I'd probably rewrite this to #ifdef bla \ cascadeIndex = Compute...() \ else \ cascadeIndex = 0 \ #endif

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Terrain/LightweightPassLitGrass.hlsl [hunk 6]
```diff
@@ -26,7 +26,7 @@ struct GrassVertexOutput
 
     half4 fogFactorAndVertexLight   : TEXCOORD5; // x: fogFactor, yzw: vertex light
 
-#ifdef _SHADOWS_ENABLED
+#ifdef _DIRECTIONAL_SHADOWS
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Terrain/LightweightPassLitGrass.hlsl]: I suppose this one deserves consistency (positionCS) as well :)

Comment (aleks01010101) [reply com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Terrain/LightweightPassLitGrass.hlsl]: ...and normalWS and viewDir?

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Terrain/LightweightPassLitTerrain.hlsl [hunk 7]
```diff
@@ -180,30 +179,26 @@ VertexOutput SplatmapVert(VertexInput v)
...
 
 #if defined(_NORMALMAP) && !defined(ENABLE_TERRAIN_PERPIXEL_NORMAL)
     float4 vertexTangent = float4(cross(float3(0, 0, 1), v.normal), 1.0);
-    OutputTangentToWorld(vertexTangent, v.normal, o.tangent.xyz, o.binormal.xyz, o.normal.xyz);
+    VertexTBN vertexTBN = GetVertexTBN(v.normal, vertexTangent);
 
-    o.normal.w = viewDir.x;
-    o.tangent.w = viewDir.y;
-    o.binormal.w = viewDir.z;
+    o.normal = half4(vertexTBN.normalWS, viewDir.x);
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Terrain/LightweightPassLitTerrain.hlsl]: these should have suffixes as well (o.normalWS and brothers)

also, you've been renaming `o` to output

File: com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightBlit.shader [hunk 8]
```diff
@@ -21,13 +21,13 @@ Shader "Hidden/LightweightPipeline/Blit"
...
+            struct Attributes
             {
                 float4 vertex   : POSITION;
                 float2 uv       : TEXCOORD0;
             };
 
-            struct VertexOutput
+            struct Varyings
             {
                 half4 pos       : SV_POSITION;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightBlit.shader]: positionCS :P

File: com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightSampling.shader [hunk 9]
```diff
@@ -8,29 +8,29 @@ Shader "Hidden/LightweightPipeline/Sampling"
...
     {
         float4 vertex   : POSITION;
         float2 texcoord : TEXCOORD0;
         UNITY_VERTEX_INPUT_INSTANCE_ID
     };
 
-    struct Interpolators
+    struct Varyings
     {
         float4  pos      : SV_POSITION;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightSampling.shader]: positionCS

File: com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightScreenSpaceShadows.shader [hunk 10]
```diff
@@ -27,61 +27,61 @@ Shader "Hidden/LightweightPipeline/ScreenSpaceShadows"
...
         {
             float4 vertex   : POSITION;
             float2 texcoord : TEXCOORD0;
             UNITY_VERTEX_INPUT_INSTANCE_ID
         };
 
-        struct Interpolators
+        struct Varyings
         {
             half4  pos      : SV_POSITION;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightScreenSpaceShadows.shader]: ...and you wouldn't believe me!
positionCS :D

File: com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightScreenSpaceShadows.shader [hunk 11]
```diff
@@ -27,61 +27,61 @@ Shader "Hidden/LightweightPipeline/ScreenSpaceShadows"
...
+            float deviceDepth = SAMPLE_DEPTH_TEXTURE(_CameraDepthTexture, sampler_CameraDepthTexture, input.texcoord.xy);
 #endif
 
 #if UNITY_REVERSED_Z
             deviceDepth = 1 - deviceDepth;
 #endif
             deviceDepth = 2 * deviceDepth - 1; //NOTE: Currently must massage depth before computing CS position.
 
-            float3 vpos = ComputeViewSpacePosition(i.texcoord.zw, deviceDepth, unity_CameraInvProjection);
+            float3 vpos = ComputeViewSpacePosition(input.texcoord.zw, deviceDepth, unity_CameraInvProjection);
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightScreenSpaceShadows.shader]: I suppose these temporaries deserve fuller names as well

File: com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightStandardUnlit.shader [hunk 12]
```diff
@@ -78,31 +79,37 @@ Shader "LightweightPipeline/Standard Unlit"
...
+                output.vertex = vertexPosition.hclipSpace;
+                output.uv0AndFogCoord.xy = TRANSFORM_TEX(input.uv, _MainTex);
+                output.uv0AndFogCoord.z = ComputeFogFactor(vertexPosition.hclipSpace.z);
 
 #if _SAMPLE_GI
-                OUTPUT_NORMAL(v, o);
-                OUTPUT_LIGHTMAP_UV(v.lightmapUV, unity_LightmapST, o.lightmapUV);
-                OUTPUT_SH(o.normal, o.vertexSH);
+                VertexTBN vertexTBN = GetVertexTBN(input.normal, input.tangent);
+                output.normal = vertexTBN.normalWS;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/LWRP/Shaders/LightweightStandardUnlit.shader]: normalWS and others :)

File: com.unity.render-pipelines.lightweight/LWRP/LightweightPipelineCore.cs [hunk 13]
```diff
@@ -109,24 +113,28 @@ void SortCameras(Camera[] cameras)
...
-            bool anyShadows = pipelineAsset.supportsDirectionalShadows || pipelineAsset.supportsLocalShadows;
+                s_ShaderFeatures |= ShaderFeatures.RealtimeDirectionalLightShadows;
+
+            if (pipelineAsset.punctualLightsSupport == RealtimeLightSupport.PerVertex)
+            {
+                s_ShaderFeatures |= ShaderFeatures.RealtimePunctualLightsVertex;
+            }
+            else if (pipelineAsset.punctualLightsSupport == RealtimeLightSupport.PerPixel)
+            {
+                s_ShaderFeatures |= ShaderFeatures.RealtimeDirectionalLights;
```
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/LWRP/LightweightPipelineCore.cs]: Thanks for catching that.

File: com.unity.render-pipelines.lightweight/LWRP/LightweightPipelineCore.cs [hunk 14]
```diff
@@ -109,24 +113,28 @@ void SortCameras(Camera[] cameras)
...
+                if (pipelineAsset.supportsPunctualShadows)
+                    s_ShaderFeatures |= ShaderFeatures.RealtimePunctualLightShadows;
+            }
+
+            bool anyShadows = pipelineAsset.supportsDirectionalShadows || pipelineAsset.supportsPunctualShadows;
```
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/LWRP/LightweightPipelineCore.cs]: I think it makes sense to check here regardless of UI. Fixed it.

File: com.unity.render-pipelines.lightweight/LWRP/Passes/SetupLightweightConstanstPass.cs [hunk 15]
```diff
@@ -66,19 +66,19 @@ public SetupLightweightConstanstPass()
...
         /// <param name="perObjectLightIndices">Buffer holding per object light indicies</param>
-        public void Setup(int maxVisibleLocalLights, ComputeBuffer perObjectLightIndices)
+        public void Setup(int maxVisiblePunctualLights, ComputeBuffer perObjectLightIndices)
         {
-            this.maxVisibleLocalLights = maxVisibleLocalLights;
+            this.maxVisiblePunctualLights = maxVisiblePunctualLights;
             this.perObjectLightIndices = perObjectLightIndices;
 
-            if (m_LightColors.Length != maxVisibleLocalLights)
+            if (m_LightColors.Length != maxVisiblePunctualLights)
```
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/LWRP/Passes/SetupLightweightConstanstPass.cs]: I'll take a look at that in a second PR as it seems these buffers should be allocated at constructor and this piece of code was not done by me. So I need to understand the reasoning.

File: com.unity.render-pipelines.lightweight/LWRP/Passes/SetupLightweightConstanstPass.cs [hunk 16]
```diff
@@ -215,48 +215,47 @@ void SetupMainLightConstants(CommandBuffer cmd, ref LightData lightData)
...
         }
 
         void SetShaderKeywords(CommandBuffer cmd, ref CameraData cameraData, ref LightData lightData, ref ShadowData shadowData)
         {
-            int vertexLightsCount = lightData.totalAdditionalLightsCount - lightData.pixelAdditionalLightsCount;
-
-            CoreUtils.SetKeyword(cmd, LightweightKeywordStrings.AdditionalLights, lightData.totalAdditionalLightsCount > 0);
-            CoreUtils.SetKeyword(cmd, LightweightKeywordStrings.MixedLightingSubtractive, m_MixedLightingSetup == MixedLightingSetup.Subtractive);
-            CoreUtils.SetKeyword(cmd, LightweightKeywordStrings.VertexLights, vertexLightsCount > 0);
+            bool realtimePunctualLightsPerPixel = lightData.punctualLightsCount > 0 && !lightData.shadePunctualLightsPerVertex;
```
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/LWRP/Passes/SetupLightweightConstanstPass.cs]: Fixed it.

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Lighting.hlsl [hunk 17]
```diff
@@ -101,76 +105,65 @@ half SpotAttenuation(half3 spotDirection, half3 lightDirection, half2 spotAttenu
...
-    lightInput.color = _AdditionalLightColor[lightIndex].rgb;
-    lightInput.distanceAndSpotAttenuation = _AdditionalLightAttenuation[lightIndex];
-    lightInput.spotDirection = _AdditionalLightSpotDir[lightIndex];
+    // Abstraction over Light input constants
+    float4 positionAndSubtractiveLightMode = _PunctualLightsPosition[perObjectLightIndex];
+    half4 distanceAndSpotAttenuation = _PunctualLightsAttenuation[perObjectLightIndex];
+    half4 spotDirection = _PunctualLightsSpotDir[perObjectLightIndex];
 
-    half4 directionAndRealtimeAttenuation = GetLightDirectionAndAttenuation(lightInput, positionWS);
+    float3 posToLightVec = positionAndSubtractiveLightMode.xyz - positionWS;
```
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Lighting.hlsl]: The way to fix this is with camera relative rendering. LW is planned to support it once we upload the objectToWorld and inverse matrix with position relative to camera.

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Input.hlsl [hunk 18]
```diff
@@ -44,15 +44,15 @@ CBUFFER_START(_LightBuffer)
 float4 _MainLightPosition;
 half4 _MainLightColor;
 
-half4 _AdditionalLightCount;
-float4 _AdditionalLightPosition[MAX_VISIBLE_LIGHTS];
-half4 _AdditionalLightColor[MAX_VISIBLE_LIGHTS];
-half4 _AdditionalLightAttenuation[MAX_VISIBLE_LIGHTS];
-half4 _AdditionalLightSpotDir[MAX_VISIBLE_LIGHTS];
+half4 _PunctualLightsCount;
+float4 _PunctualLightsPosition[MAX_VISIBLE_LIGHTS];
```
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Input.hlsl]: Fixed it. Turns out the subtractive attenuation light mode is only used in the main directional light so we could simplify a few bits here.

File: com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Lighting.hlsl [hunk 19]
```diff
@@ -101,76 +105,65 @@ half SpotAttenuation(half3 spotDirection, half3 lightDirection, half2 spotAttenu
...
+    float distanceSqr = max(dot(posToLightVec, posToLightVec), HALF_MIN);
+
+    half3 lightDirection = half3(posToLightVec * rsqrt(distanceSqr));
+    half attenuation = DistanceAttenuation(distanceSqr, distanceAndSpotAttenuation.xy);
+    attenuation *= AngleAttenuation(spotDirection.xyz, lightDirection, distanceAndSpotAttenuation.zw);
```
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/LWRP/ShaderLibrary/Lighting.hlsl]: We can't classify these light types on forward. Branching is not possible, adding separate loops for point and spot is not ideal as well. AngleAttenuation is done with 3 simple ALU (dot, mad, scalar multiply). Most likely both distance and angle can fit into one gpu cycle in Midgard arch.

**Pull request merged**